### ISSUE TITLE
Lua/stdlib: global vs local "enable" behavior

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#28603]
### Problem

`vim.diagnostic.enable(nil)`: https://github.com/neovim/neovim/blob/e778e0116198470ba037b9426f4ff7fa5cb7f880/runtime/lua/vim/diagnostic.lua#L2046-L2053 and `vim.lsp.inlay_hint.enable(nil)`: https://github.com/neovim/neovim/pull/28543

both actively override any buffer-local flags that were previously set by `enable(..., {bufnr=...}`. 

That seems a bit "destructive". I tend to think that global enable() should act like a gate: 

- if globally enabled, then defer to the buffer-local flag
- if globally disabled, don't check the buffer-local flag

IOW: buffer-local "enable" flag has precedence unless the feature is globally disabled.

### Expected behavior

Need a clear pattern for all such modules to follow. This may have implications for other kinds of "config" besides mere "enablement".

The behavior of vim *options* (`:help global-local`, ignoring other kinds of option scopes) would imply that the global setting is only the "default", and any buffer-local setting should always override the global default.

**Comment by ofseed** [comment]
IMO the current behavior of `enable` is good enough, even if we provide something like `vim.g.inlay_hint = true` and `vim.b.inlay_hint = false`, we should keep the current behavior.

A summary of the current pattern:

**Keep buffer-local states only when it is different from the global one**. This means that after it is equalized, its state will change as the global state changes.

As I said in https://github.com/neovim/neovim/pull/28543#discussion_r1587049857:

>The logic behind this design is that if it can be globally `enable` or locally `enable`, when the user selects local `enable`, we think the user's purpose is to distinguish the local state from the global state. I think this design is good for interaction,

It is the best choice that can be made under the following constraints:

>...there is no way to remove "local variables" by using `enable`. Once the user manages a buffer's `bufstate`, it means that they must manage it all the time, there's no way to make them follow the `globalstate`.

**Comment by mehalter** [comment]
I agree with ofseed here regarding global and local states. If we adopt a model where the global state acts as a gate and then defer to local states, the way of modeling "I want to have this feature disabled but I want to enable it for a single buffer" becomes very difficult to represent. If the user has an option disabled globally and they do something to toggle locally then nothing would happen. Then if they enable it globally it could end up applying to all open buffers even if the user doesn't mean to do that.

I tend to like that the current design logic represents, "we do the most specific thing that the user has asked for so far"

**Comment by ofseed** [comment]
It might be time to decide on this behavior; at least I can consider the current behavior (that is, what `vim.lsp.inlay_hint` does) is acceptable, right? Because this can be handled in the `vim.lsp._state` that I am about to try to implement, avoiding the need to write the logic to check if it's enabled everywhere.

**Comment by justinmk** [comment]
Looks like there is not much demand for anything different than what exists, so I agree with your comments above, and then the next thing is to write down the conclusion.

Also, in general, if a builtin module needs a mature "options" behavior, it should... graduate its quasi-options into true editor options. That will be helped by https://github.com/neovim/neovim/issues/25672 . Perhaps eventually we could support "namespaced" options, so plugins could define their own subtree of options and get the full options capabilities (scoping, set vs setlocal, etc.).

**Comment by lewis6991** [comment]
We may need something similar for `vim.lsp.config` and `vim.lsp.enable` to better accommodate project specific setups.
